CIP-0118/README.md

```diff
@@ -0,0 +1,1042 @@
+in the ledger state, which we call the FRxO.
+
+**Checking minting policies**.
+In the original paper, we attempted to prevent flash loan attacks by forcing
+the execution of minting policies for any assets for which the total negative or
+positive quantities were changed. This was not a viable strategy for existing
+tokens, for which minting policies would overwhelmingly be unequipped to reason
+about such cases. In this proposal, we instead perform the cycle detection check
+on the transaction dependency graph.
+
+
+### Westberg’s “Smart Transactions”
+
+The design [Smart Transactions](https://github.com/input-output-hk/Developer-Experience-working-group/issues/47)
+has a lot of similarities to ours, in particular it follows the approach where
+transactions have partially unspecified inputs and outputs.
+However, it goes much further and proposes conditions on how those can be satisfied,
+as well as some kind of computed conditions for updating datums.
+It also proposes that intent-processing be done by the node.
+
+We think that we can eventually move in the direction of adding more of the
+intent-settlement features from the proposal (e.g. conditions on values in inputs),
+but the major point of disagreement is about intent processing.
+As previously discussed, this proposal only focuses on intent settlement, not intent processing.
+We now briefly discuss options of how to possibly proceed in this direction.
+
+
+### Validation zones for settling other types of intents
+
+This CIP lays the foundation for specification and resolution of intents on Cardano.
+It does so by establishing the concept of a validation zone - a list of transactions that
+can be validated as a whole, but it is impossible to check if the individual transactions
+are valid outside a zone. We conjecture that infrastructure we propose can be used for processing
+other types of underspecified transactions, and therefore more sophisticated
+intents. A proof of this requires a
+fleshed-out design for the expanded intent-processing system, which is outside the
+scope of this CIP. However, possibilities include :
+
+- Intent transactions that allow output references to be removed and replaced with their corresponding
+output data, while intent-settling transactions provide the matching output references
+for those outputs to fulfill this intent. This is the ledger component of a proposal
+for a light client solution.
+
+- Intents that support more advanced kinds of swaps, such as limit or market orders,
+and do not require the entire quantity offered to be swapped at once.
+
+Supporting these sorts of intents require the definition of new types of temporary ledger
+state structures, as well as changes to zone-level constraints.
+We believe that this is a fruitful direction that we can extend in future, and,
+as such, we think we can progress with this design even if we don’t have a full
+picture of where we might want to go with intents, although this is a significant risk.
```

Comment:
@polinavino I really appreciate the R&D!

IIUC my biggest issue is the absence of partial fulfillment (ie, the values must match exactly). I think the absence of this feature severely minimizes its economic utility since the double coincidence of wants is likely to be a huge obstacle for most economic applications that try to use these Validation Zones.

> ... we think we can progress with this design even if we don't have a full picture of where we might want to go with intents, although this is a significant risk.

I don't agree with this stance. In my view, the minimum viable product of an intent system *must* support partial fulfillment. I am definitely no where near as well versed with the Validation Zone design as you, but it is not clear to me that the current design actually *can* be cleanly extended to support partial fulfillment. I personally don't find the other use cases compelling enough to take on this risk since they seem too niche. Even the DEX aggregator example only really seems relevant for liquidity pool based DEXs.

I would totally be in favor of this if I had more assurances that Validation Zones could be extended to support partial fulfillment.

Reply:
Other than minUTXO concerns, couldn't partial fulfilment be handled by an extra output, and even an extra transaction to merge those outputs?

Reply:
I'm not entirely sure what you mean. Where would the value for the extra output come from? I'm imagining a third-party aggregator trying to settle a big limit order intent against a bunch of smaller intents. An example would be a limit order looking to buy 10,000 ADA being settled against a bunch of smaller limit orders that are only offering 3,000 ADA in total. If a third-party aggregator wanted to batch these orders together using Validation Zones, the aggregator would need to front the difference of 7,000 ADA so that the big limit order intent can actually be used. This is impractical for all, but the most wealthy third-party aggregators. This isn't even an aggregator anymore since fronting the ADA actually makes it a market maker.

Reply:
Limit orders wouldn't be the ones taking advantage of this, because they're long lived and so can afford the contention; they would be implemented as a UTXO on chain that ensures the user is paid their limit price.

It's market orders that would utilize this; the user would construct a transaction in which the token they are selling doesn't appear in the outputs, but the minimum token they're ok receiving is paid to their address.

An off-chain process then matches them against limit orders, other market orders, or AMM liquidity pools in a validation zone; the minimum output is satisfied by whatever liquidity source, because the user doesn't care where it comes from, and if the tx fails because of contention, they can retry and the users signed tx is still valid.

The limit order contract allows partial matches, enforcing that any surplus is paid back into the contract. Since it's sitting there long term, it can afford the contention: it's other people that are spending it, and it doesn't need to provide a signature.

Any surplus over the market orders minimum can be sent to the user in a separate output, barring any minutxo issues, hence my comment above.

Reply:
> Limit orders wouldn't be the ones taking advantage of this, because they're long lived ...

Not all limit orders are long lived. Consider day-traders who are waiting to see if their prices get hit. They can update their limit orders several times within an hour. It would be better if this activity happened off-chain.

> An off-chain process then matches them against limit orders, other market orders, or AMM liquidity pools in a validation zone ...

Perhaps I am misunderstanding something, but AFAIU this is not practical because you can only match against outputs whose values match the off-chain orders *exactly*. This is the double coincidence of wants problem. What are the odds that you can find a limit order or other market order that exactly matches your off-chain order? I believe you even [described](https://github.com/cardano-foundation/CIPs/pull/780#issuecomment-1992362771) in the other PR's discussion that AMMs would first need to break off smaller outputs that then can exactly match the off-chain orders. No DApps currently support "breaking off" outputs, and I am against adding this feature since it involves dramatically increasing the DApp's complexity and attack surface just to make up for a shortfall in the Validation Zone design.

> The limit order contract allows partial matches, enforcing that any surplus is paid back into the contract.

AFAIU partially matching on-chain limit orders against off-chain orders is only possible if the limit order contract supports breaking up the limit order into the fulling output(s) and change output when it needs to be matched against off-chain orders. As I said, I'm against DApps adding support for this.

Reply:
Ok, well, feel free to design things however you want 😅

I was just pushing back agains the claim that the minimum viable design *must* include partial fulfillment. I'm asserting that this design is substantially useful already, and thus serves as a suitable minimum viable design.

Reply:
It is true that this design is intended for the case where the initial submitter "needs a certain amount of currency X, and is not ok with getting a lesser amount". The original purpose of this project was paying transaction fees in non-Ada tokens. 

I think it might be possible to support more true DEX-like function, but it would require some extra off-chain communication and complexity. The idea is that users must submit requests alongside the requested amounts, to show that they can cover any funds in the requests that were not exchanged for their desired currency. Then they have to spend some of those funds to cover their own un-exchanged requests. Consider the following change to the DEX aggregator usecase :

1. Parties A1 ... An create transactions T1 ... Tn with requests and offers (unlocked outputs) representing their desired trade, as in the previous example.
              - *here, users actually include the amount they request in the inputs AND the outputs AND the requests* 
              - e.g. a request with 3 PolinaCoins must be in a transaction with 3 inputs containing 1 PolinaCoin and 1 output containing 3 PolinaCoins
              
2. Parties A1 ... An send T1 ... Tn to B

3. Party B creates two transactions:
    - T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. It also requires T(n+2)
             - *at this point, there might still be some tokens missing from the zone, if not all requests have been fully covered by zone transaction, so there are some requests in T(n+1) still. The exchange must keep track of which user's requests are being only partially fulfilled* 
    - T(n+2) spends the offers from T1 ... Tn as well as `u'`, creating a new increased liquidity output `u''`
    
4. B sends the unfulfilled requests (in the form of a UTxO set) to the corresponding users , and gets them to sign transactions T1' T2' which (i) spend the missing requests that remain after the aggregation (and `requireTxs` the transactions in (ii)), and (ii) spend the outputs containing the requested amounts (e.g. 3PolinaCoin output above) to cover the missing requested amounts

5. All parties check and sign the transactions from B that "refund" the unexchanged funds, send them back to B

6. Party B submits a validation zone consisting of T1 ... T(n+2) ; T'1 ; T'2

This is not fully developed usecase yet, but the point is that it uses this infrastructure to to do something that works like partially fulfilling requests.

Reply:
> 4. B sends the unfulfilled requests (in the form of a UTxO set) to the corresponding users , and gets them to sign transactions T1' T2' ...
> 5. All parties check and sign the transactions from B that "refund" the unexchanged funds, send them back to B

IIUC these step introduce a user coordination problem that I think makes this impractical at scale, and will actually create perverse incentives for the aggregator. 

Consider if Alice, Bob, and Charlie all submit off-chain orders as you describe and an aggregator batches them together using Validation Zones. In this batch, all user orders are only partially fulfilled. The aggregator must send the required refund transactions to each user, and wait for them to sign and return the transactions before it can submit the batch.

What if Alice and Bob quickly sign their required refund transactions, but Charlie is unavailable? This would likely mean UTxOs being spent by the batch will be taken by other aggregators/users and this batch transaction is now invalid. Alice and Bob are therefore hurt by Charlie's slow response. This problem will likely happen frequently, especially if this is a global marketplace where all parties are strangers and live in separate time zones. In practice, this means Alice, Bob, and Charlie will need to always be online and quickly sign all refund transactions to maximize the chance of at least one going through. Most of these transactions will likely fail due to the coordination problem. This would be a terrible user experience, and is probably bad enough to deter most people from using this system in the first place.

This is also bad from the aggregator perspective since the aggregator likely only gets paid if the batched transactions are successfully submitted. The more partial users the aggregator tries to batch together, the more the coordination problem will be an issue. To minimize this risk, aggregators will either batch the minimum number of partial users required, or simply avoid any partial filling of orders to avoid the coordination problem entirely.

Reply:
Those are all fair points.

What it seems like you really need for correctly handling the usecase you propose is something like a "ledger primitive" for a DEX contract, alongside multi-sig and Plutus. That is, a language wherein the only thing you can express is "the tokens in this output are for anyone to take so long as they send some tokens at the specified rate r to my address a", or something like that. All the matching would be done off-chain, eg. by an aggregator. If designed as an extension of the Validation Zones proposal, this would have the added benefit (in addition to cheaper script costs) that transactions with offers do not need to depend on each other or be in particular sequence, so long as all the money in the offers adds up. A transaction that "closes out the zone by paying out the un-exchanged funds" is still needed, but can be made by anyone (no specific signatures required). 

The purposes of these two designs are different enough that they would constitute two different kinds of intents - one where the entire expected payout from an offer can be viewed as funds available for use inside the transaction, and one where the expected payout is not available inside the transaction. It may or may not be possible to generalize these two - as well as all future types of intents - as some unified feature, but this is far beyond the current research being done. Adding different intents as separate features using Validation Zones, however, is much more accessible, especially if we already have evidence that this validation zones usecase (babel fees) works.

Reply:
The point I am making here is that we consider this CIP as extensible for other kinds of intents, and we would like to asses the response to the Validation Zones design specifically for the case of Babel fees before we introduce more kinds of intents are essentially non-overlapping. 

Would you consider satisfactory a detailed "next steps" section where we discuss a possible design to implement what you are requesting (this would likely not include additional code changes or attack vector analysis, only a specification)?

Reply:
I think I actually came up with an approach to partial matching that doesn't require any changes to the Validation Zone design and also doesn't require "breaking off" pieces of on-chain liquidity sources. AFAICT all of the approaches described so far have been from the perspective of matching each off-chain intent against an on-chain liquidity source. However, it may be more practical to match off-chain intents against each other and *only match the value difference against an on-chain liquidity source*. This approach has the benefit of being able to "on-load" partial fulfillment to the on-chain liquidity sources without having to break up the on-chain liquidity UTxOs.

For a simple example, imagine if Alice has 10 ADA and wants to convert it to 9 DJED. She already has the 10 ADA sitting in a UTxO so she just needs to create the intent transaction, sign it, and send it to an aggregator. This UTxO with 10 ADA will be the fulfilling output for her intent transaction.

Meanwhile, Bob has 8 DJED and wants to convert it to 9 ADA. Like Alice, he already has the 8 DJED in a UTxO so he can just create the intent transaction, sign it, and send it to an aggregator. This UTxO with 8 DJED will be the fulfilling output for his intent transaction.

The aggregator sees the two intent transactions, but can't match them because the values don't match exactly: there is a net value difference of +1 ADA (10 ADA input - 9 ADA output) and -1 DJED (8 DJED input - 9 DJED output). This is the double coincidence of wants problem, but it is easily solvable by using an on-chain liquidity source to convert the extra 1 ADA to 1 DJED. On-chain liquidity sources can be partially fulfilled so it would be easy to find a liquidity source to use.

So the aggregator matches the two intents against each other and matches the value difference against the on-chain liquidity source to produce a valid transaction. The required fulfilling outputs for the intent transactions *already exist* so there is no need for the aggregator to try breaking off pieces of the on-chain liquidity UTxO.

I think this approach works especially well for on-chain liquidity sources that can be composed. For example, if Alice wanted to convert ADA -> WMT and Bob wanted to convert AGIX -> DJED, these two off-chain intents can't be directly matched against each other. However, if there is a liquidity source for WMT -> USDC and another for USDC -> AGIX, the off-chain intents can be indirectly matched together as long as the on-chain liquidity sources can be composed to effectively create a WMT -> AGIX conversion. So this transaction would have 2 off-chain intents + 2 on-chain liquidity sources. It is even possible to batch different pairings of intents+sources as long as the liquidity sources can be composed in the same transaction.

From an incentives perspective, aggregators will actually want to do this, even without charging a fee, because the possible compositions enable finding huge arbitrage opportunities. It is possible for no collateral to ever be at risk with this approach (it depends on the liquidity source logic). I explained these incentives in-depth in my cardano-swaps DEX [readme](https://github.com/fallen-icarus/cardano-swaps?tab=readme-ov-file#naturally-incentivized-liquidity) which uses the same arbitrage possibilities to incentivize aggregators.

This approach also enables users to submit limit orders off-chain, and allows them to update their asking prices off-chain as well (the day-trading use case). The user would need to trust the aggregator to use the newest intent transaction, but all of the cancellation methods mentioned in the CIP would still work with this method so users can revoke permission at any time.

@polinavino Do you have any thoughts on this? Did I misunderstand something that actually makes the above impossible?

Reply:
Let me try to understand this better - it sounds to me that the following assumptions would be required for what you are suggesting :

- all transactions involved in an exchange of this sort (i.e. contained in the zone) have to be exchanging only the same two types of assets, and they have the same exchange rate between the two. In your example, the exchange rates are different, so I don't quite see why this is ok - but it would work for Alice offering 10DJED for 5Ada, and Bob offering 4Ada for 8DJED.

- the liquidity pools basically absorbs any left over un-exchanged assets by covering any remaining requested tokens and appropriating the extras. In the case I described, DJED pool would be getting 2 DJED and Ada pool would be losing 1Ada

Does this sound right at all?

Reply:
> all transactions involved in an exchange of this sort (i.e. contained in the zone) have to be exchanging only the same two types of assets, and they have the same exchange rate between the two. 

No, both the types of assets and the exchange rates are ultimately irrelevant. AFAIU the only requirements for validation zones are:
- All input "holes" must have a corresponding fulfilling output with the proper value that comes from the proper address.
- The total value of the validation zone must be balanced.

In my simple example, Alice had a UTxO with 10 ADA and created an intent transaction with an input "hole" for 10 ADA from her address, and an output to her address with 9 DJED. To actually execute Alice's intent, the aggregator just needs to specify Alice's UTxO with 10 ADA as the fulfilling output for her intent transaction. The same applies to Bob's intent transaction.

So the validation zone containing both Alice's and Bob's intent transactions has all input "holes" satisfied, but is still unbalanced overall. To balance it, just add another normal input. You don't need to use another intent transaction to finish balancing the zone; a normal transaction, without any input holes, would work. So the validation zone consists of:
- 2 intent transactions and 1 normal transaction
- Alice's 10 ADA UTxO and Bob's 8 DJED UTxO are specified as fulfilling outputs for their respective intent transactions.

All input "holes" are filled and the value of the overall validation zone is balanced.

> - the liquidity pools basically absorbs any left over un-exchanged assets by covering any remaining requested tokens and appropriating the extras. In the case I described, DJED pool would be getting 2 DJED and Ada pool would be losing 1Ada

This part is correct.

Reply:
ok, i think i see - let me try again - 

There is actually no change to the DEX aggregator example (below), we just **do not make the assumption that the total liquidity in B's pool stays constant**, and instead assume that B participates in the exchange (exchanging some of their own assets) in the case the requests and offers of T1... Tn do not balance out perfectly.

1. Parties A1 ... An create transactions T1 ... Tn with requests and offers (unlocked outputs) representing their desired trade, as in the previous example.
2. Parties A1 ... An send T1 ... Tn to B
3. Party B creates two transactions:
    - T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. It also requires T(n+2)
    - T(n+2) spends the offers from T1 ... Tn, placing it into liquidity output `u''`
4. Party B submits a validation zone consisting of T1 ... T(n+2)
    - Possibly followed by a transaction aggregating liquidity in `u'` and `u''`

Reply:
> we just do not make the assumption that the total liquidity in B's pool stays constant, and instead assume that B participates in the exchange (exchanging some of their own assets) in the case the requests and offers of T1... Tn do not balance out perfectly.

This part is correct.

> 3. Party B creates two transactions:
    - T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. It also requires T(n+2)
    - T(n+2) spends the offers from T1 ... Tn, placing it into liquidity output `u''`

This part is incorrect. Only one transaction is required from Party B. T(n+1) uses `u` to finish balancing T1 ... Tn and produces `u'` as a result; this is where the absorption happens. Going from `u` to `u'` required depositing the un-exchanged assets and withdrawing the required assets to finish balancing T1 ... Tn. It stops there. There is no T(n+2) and no `u''`. The validation zone consists of T1 ... Tn and T(n+1). If liquidity sources can be composed, T(n+1) can contain multiple sources that each absorb different parts of the left over un-exchanged assets.

>  T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. 

Specifically, this part is wrong. The fulfilling UTxOs already exist so there is no need to break off chunks from `u` to fulfill the requests in T1 ... Tn.

Reply:
> The reason there must be two transactions is that it is not possible for a transaction to spend both a regular output and a request from any other transaction.

Does this mean you cannot use a "pre-existing" UTxO to fulfill a request? You must first create the UTxOs in a transaction within the zone for it to be used as a fulfilling UTxO?

For example, is this not possible:
1. Parties A1 ... An create transactions T1 ... Tn with requests and offers (unlocked outputs) representing their desired trade, as in the previous example. The UTxOs meant to fulfill the requests already exist at each parties' address.
2. Parties A1 ... An send T1 ... Tn to B
3. Party B creates two transactions:
    - T(n+1) fulfills all of the request by spending the required UTxOs (that already exist) from Parties A1 ... An.
    - T(n+2) spends liquidity source `u` to finish balancing the zone. This results in `u'` being created.
4. Party B submits the zone consisting of T1 ... T(n+2).

T(n+1) matches the pre-existing fulfilling outputs to the requests for them to be used in T(n+2).

Reply:
> Does this mean you cannot use a "pre-existing" UTxO to fulfill a request? You must first create the UTxOs in a transaction within the zone for it to be used as a fulfilling UTxO?

It does not mean that, no. It means that if a transaction `tx1` fulfills a request `r` made by transaction `tx2`, it *has* to use a UTxO that is *not* created by `tx2` or any transaction on which `tx2` depends, so that the liquidity used to fulfill `r` could not possibly come from "thin air", i.e. splitting `0` assets into `r` and `-r` assets, and then re-combining these into `0`. 

> T(n+1) fulfills all of the request by spending the required UTxOs (that already exist) from Parties A1 ... An.

This approach requires the "UTxOs (that already exist) from Parties A1 ... An" to be locked by contracts such that they can only be spent if requests are also being fulfilled in the same transaction, which kind of defeats the purpose of using this feature, since the trade is basically done with a contract at that point (which requires more coordination).

In the T(n+1) and T(n+2) i described above, all offers made by A1...An are unlocked. The idea is that even though they are unlocked, the transactions T1...Tn cannot be applied without some other transaction(s) covering the missing value in the requests (which is what T(n+1) does, and also requires T(n+2) to be present), and then T(n+2) collects all the unlocked outputs. No scripts required. Although two transactions are needed, we anticipate that the difference in transaction size between somehow allowing 1 transaction to both fulfill requests and collect offers, and the 2-transaction design we present would likely be O(1), since it is sort of like splitting one transaction into 2, but both parts would be signed separately.

Reply:
@polinavino I appreciate you continuing to answer my questions regarding this CIP! It seems I've been misunderstanding how intents are ultimately modeled in Validation Zones. The more I learn about Validation Zones, the more its design seems to diverge from the way I've always thought Cardano intents would work.

> ### Open (atomic) swaps
>
> A wants to swap 10 Ada for 5 tokens `myT`.
>
> 1. Party `A` creates a transaction `tx1` with
>    - an unlocked request `0` with 5 `myT` tokens,
>    - some inputs totalling 10 Ada
>    - an output `0` which contains 5 `myT` tokens, and
>    - an unlocked output `1` containing 10 Ada
> 
> -- *reordered to group inputs with outputs*

Validation Zones model intents as *completed transactions*, albeit with some holes to fill (the CIP calls these "underspecified transactions"). The above transaction (`tx1`) is fully balanced and specifies *exactly* what each input/output should look like:
- only one extra input is allowed and it must contain 5 `myT`
- only one extra output is allowed and it must contain 10 Ada

But this is not what an intent *actually* is. An intent is just a *piece* of a transaction. This is perfectly exemplified by the above descriptor: `A wants to swap 10 Ada for 5 tokens myT`. This doesn't say anything about the other pieces of the transaction. In the atomic swap above, the intent is more accurately modeled by just:
> - some inputs totalling 10 Ada
> - an output `0` which contains  5 `myT` tokens

The other unlocked input (aka request) and unlocked output in `tx1` do nothing but restrict how this intent can be used:
- You cannot combine intents into a single transaction since they are already complete transactions. Instead, Validation Zones enables batching transactions but this adds cost to the aggregator since each transaction has a minimum 0.15 ADA fee. The fact that all Validation Zone intents are separate transactions means this cost can quickly add up.
- You cannot perform other actions in the intent transaction, even if the intent owner is actually okay with it. If Alice just wants the above atomic swap, she doesn't care if the aggregator also satisfies someone else's intent in the same transaction. The aggregator could also do a swap of their own and Alice still wouldn't care. Forcing each intent to be an isolated transaction seems like a waste of space and money.
- The intent owner has to *guess* what the missing inputs and outputs would look like and place these holes in the transaction. If they guess wrong, this places enormous friction (aka, increased cost) on how an aggregator can actually fulfill this intent. Specifically, the aggregator must contort (ie, break up or merge) other UTxOs to fit the holes set in each intent. The batched transactions in Validation Zones is also meant to address this, but this means the previous bullets are exacerbated since each Validation Zone intent actually requires *multiple* transactions to fulfill it.

I think the eUTxO model actually enables modeling intents more closely to what they are: *pieces of transactions*. Alice can create a piece of a transaction, sign it, and send it to the aggregator. Bob can do the same. The aggregator can take these two transaction pieces and fit them together monoidally (ie, like a puzzle) to create a full transaction. I am making an important assumption here: **a collection of *signed* transaction pieces is just as valid as a single transaction signed by everyone**. I don't see how aggregating the pieces would invalidate the signatures. The fact that each UTxO is atomic seems to guarantee each piece can be considered an *independent event*. This is effectively a multisig where each party only needs to sign the parts of the transaction that are actually relevant to them. It is a *decentralized multisig*.

An empty transaction piece is effectively just a transaction with all fields empty. Users can add pieces they care about and leave the rest blank. So for the above atomic swap, Alice can create a transaction piece with one input from her address for 10 ADA and one output to her address for 5 `myT`; everything else is left empty. Bob can create his intent with one input, 2 outputs, and a validity interval; everything else is left blank. As long as there are no conflicts between the pieces, these two pieces can be fit together to create a transaction with 2 inputs, 3 outputs, and the validity interval from Bob's intent. The aggregator can even add their own piece to the mix to finish balancing it. If Alice and Bob each signed their respective pieces, why does it matter what else happens in the transaction? Alice's input is still only spendable if she gets her output in the same transaction. 

When the completed transaction is submitted, the ledger would verify the transaction is a valid combination of all of the pieces and all pieces are properly witnessed. The pieces and their signatures would need to be somewhere in the transaction. Since the overall transaction can be derived from the pieces, the CDDL can be redesigned around this idea.

For smart contracts, users can specify mints, observers, staking scripts, etc, in there respective pieces. However, unlike how pubkey signatures only need to sign their respective pieces, the smart contracts would be executed against the aggregated (aka completed) transaction. This means smart contracts can even be used to *disallow* certain combinations of intent pieces.

Cancellation would work the same way as in Validation Zones. Just include a reference input or normal input in your transaction piece. To cancel it, just spend the input. These transaction pieces can even contain input "holes" (ie, unspecified output references for certain inputs). I personally don't see a need for output holes; they seem to have been introduced into Validation Zones just to compensate for its restrictiveness.
 
This approach even seems much cleaner to use for Babel Fees. For example, imagine if Alice wants to buy a sword NFT worth 25 GEMS and the transaction fee is worth about 1.4 GEMS (total of 26.4 GEMS). If Alice has a UTxO with 30 GEMS, she can create an intent piece that only specifies an input from her address for 30 GEMS and an output to her address for 3.6 GEMS + the sword NFT; all other parts are left empty. She would sign this piece and send it to the game devs who can now batch Alice's intent piece with other users who also want to purchase game items. Since these intent pieces can easily be combined, it is possible for the game devs to sell a variety of game NFTs to multiple, unrelated users in a single transaction. Doing the same thing with Validation Zones seems overly complicated and expensive (many transactions are required).

I'm not sure of everything that would need to change to support transaction pieces, but on the surface it seems way more expressive and more cost effective than Validation Zones. It doesn't seem like anyone has considered breaking up transactions and signing the pieces like this. I recognize this is probably a major change, but the expressiveness may make it entirely worth it. Is there some reason why it wouldn't work?

@Quantumplation you and I seem to disagree on everything :smile:. Do you have any criticisms of the intent piece concept?

Reply:
I don't have the spare cycles to digest it all, but I don't have any particular problems with it. But, it's just a different system than what's being proposed here. It feels a bit like playing Starcraft 2 and saying "man, this would be great if it was first person and had guns!" 😅

I think what is actually being proposed is powerful enough (or can be extended to support) all the same use cases; just because it doesn't fit what you were picturing doesn't mean it's a deficient design.

I think the main thing that this one provides that yours doesn't (and I'd have to think it through to digest this fully) is probably related to keeping transactions deterministic, and not completely rewriting all of Plutus 😅

Reply:
> I don't have the spare cycles to digest it all, but I don't have any particular problems with it.

All good. I tend to write a lot :sweat_smile:. As long as you think it is reasonable, that is enough for me.

This CIP mentions the following:

> It may be that the best way to implement intents on Cardano is entirely different from the underspecified transactions and validation zones approach we are taking here. In that case such a design might entirely obsolete this one, making it an expensive and pointless burden to maintain in future.

I just want to make sure all possible paths have been considered to avoid this burden.

> I think the main thing that this one provides that yours doesn't (and I'd have to think it through to digest this fully) is probably related to keeping transactions deterministic, and not completely rewriting all of Plutus 😅

AFAIU eUTxO guarantees the determinism even if broken into pieces, but I totally recognize this is currently just an assumption on my part. Even for ordering of the pieces, the pieces can be applied in the order they appear in the list in the CDDL (which can possibly be configurable).

Reply:
> You cannot combine intents into a single transaction since they are already complete transactions. Instead, Validation Zones enables batching transactions but this adds cost to the aggregator since each transaction has a minimum 0.15 ADA fee. The fact that all Validation Zone intents are separate transactions means this cost can quickly add up.

This is true. Since zones are limited in size to the size of a single transaction, we should consider charging 0.15 ADA per zone (or perhaps slightly more, but not like, 0.15 PER transaction). I will bring this up with the ledger team. 

> You cannot perform other actions in the intent transaction, even if the intent owner is actually okay with it. If Alice just wants the above atomic swap, she doesn't care if the aggregator also satisfies someone else's intent in the same transaction. The aggregator could also do a swap of their own and Alice still wouldn't care. Forcing each intent to be an isolated transaction seems like a waste of space and money.

I dont really get this. You can do whatever you want inside a zone, except use old-version Plutus scripts. For regular atomic swaps, no aggregator is needed, it is just between two parties. The point of designing Validation Zones in the way we did is to reduce the off-chain communication required between parties that want to perform a transaction. None at all is required - you construct a transaction in which you say what tokens you want, and what tokens you are offering in exchange, and a counterparty (you dont care who it is) builds the fulfillment, submitting the complete zone to the network. This is definitely more streamlined than co-constructing a transaction that everyone needs to sign, or a zone that everyone needs to sign. No need for the use of expensive contracts, either. Moreover, the offer and the acceptance of it is done in the same block. You are right in that it might be more expensive because of the number of transactions, however, we may be able to reduce the per-transaction cost. 

I think, actually, all of what you are suggesting is already possible in the current model using co-constructed transactions and smart contracts? 

> The intent owner has to guess what the missing inputs and outputs would look like and place these holes in the transaction. If they guess wrong, this places enormous friction (aka, increased cost) on how an aggregator can actually fulfill this intent. Specifically, the aggregator must contort (ie, break up or merge) other UTxOs to fit the holes set in each intent. The batched transactions in Validation Zones is also meant to address this, but this means the previous bullets are exacerbated since each Validation Zone intent actually requires multiple transactions to fulfill it.

Non-communicating parties that want to perform some exchanges already have to construct individual transactions to put their offers on-chain. I am not sure if multi-party co-constructing of a single transaction that includes their offers and liquidity from an exchange is current practice, but it sound cumbersome, and would still be a big expensive transaction that takes a lot of off-chain communication. In our design, everyone constructs a single small transaction and sends it off. The exchange then constructs two (big) transactions, spending everyones offers and requests. Exchanging lots of peoples offers will always require a transaction with lots of UTxOs (at least one per user).

Reply:
@fallen-icarus 
> AFAIU eUTxO guarantees the determinism even if broken into pieces,

Not true, almost by definition.

First, the assembler has freedom to choose in which order to assemble the pieces; the order of the transaction outputs, for example, can be changed. Additionally, it's impossible to construct the script context to evaluate transactions, and so whoever provides collateral might be putting their collateral at risk, for example. I'm sure there are half a dozen other subtleties that make this a nontrivial thing to spec out a *dramatic* (almost complete) change to the ledger rules / execution model for plutus scripts 😅